lsp-mode.el

```diff
@@ -5203,18 +5238,52 @@ It will show up only if current point has signature help."
         :context `( :diagnostics ,(lsp-cur-line-diagnostics)
                     ,@(when kind (list :only (vector kind))))))
 
-(defun lsp-code-actions-at-point (&optional kind)
-  "Retrieve the code actions for the active region or the current line."
-  (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind)))
+(defun lsp-code-actions-at-point (&optional kind enabled)
+  "Retrieve the code actions for the active region or the current line.
+If ENABLED is specified, only return code actions that are not
+disabled."
+  (--doto (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind))
+    (when enabled
+      (cl-callf2 seq-remove #'lsp:code-action-disabled? it))))
 
 (defun lsp-execute-code-action-by-kind (command-kind)
   "Execute code action by name."
-  (if-let ((action (->> (lsp-get-or-calculate-code-actions command-kind)
-                        (-filter (-lambda ((&CodeAction :kind?))
-                                   (and kind? (equal command-kind kind?))))
-                        lsp--select-action)))
-      (lsp-execute-code-action action)
-    (signal 'lsp-no-code-actions '(command-kind))))
+  (let* ((all-actions (->> (lsp-get-or-calculate-code-actions command-kind)
+                           (-filter (-lambda ((&CodeAction :kind?))
+                                      (and kind? (equal command-kind kind?))))))
+         (enabled-actions (-remove #'lsp:code-action-disabled? all-actions))
+         (action (lsp--select-action (or enabled-actions all-actions))))
+    (lsp-execute-code-action action)))
+
+(defun lsp-execute-code-action-by-type (kind &optional enabled)
```

Comment:
The difference between this and `lsp-execute-code-action-by-kind` isn't clear to me (or from the docs).

Reply:
`lsp-execute-code-action-by-type` matches base-types, i.e. `lsp-execute-code-action-by-type "source"` would match "source.organizeImports", while `-by-kind` matches exactly, i.e. the same example would yield nothing, and it would have to be `(lsp-execute-code-action-by-kind "source.organizeImports")`.

Reply:
I'm not sure this is a distinction the spec really knows about - certainly "base type" doesn't appear. They do talk about "base kinds", but I'm not sure they're meaningfully distinct from "kinds", it just indicates that they should be hierarchical parents of a set of kinds.

What about just having an `exact-match` argument to `lsp-execute-code-action-by-kind`? Or maybe we shouldn't even provide exact matching. The intention for spec methods that allow you to do kind-based filtering is to include kinds "under" the selected kinds (see e.g. https://github.com/microsoft/language-server-protocol/issues/970#issuecomment-633938017). Do we have a usecase for exact matching?